subsys: samples: doc: arch: Make perf subsystem

KushnerovMikhail opened the pull request:
Add profiling util based on periodic stack unwinding. Perf from Linux was taken as a reference.

Code is based on Sampling profiler (#29304) by @Jongy.

The operation of module is based on frame pointer usage and saving registers during interruption handling by zephyr.

General work description:
The unwinding function stay in timer as expiry function so is called during interruption handling. Thus the function have access to saved registers (program counter and frame pointer in particular) of the current thread and use it to unwind the thread stack.
Timer starting and results printing function are made as shell commands for conveniency.
Obtained samples can be translated into flamegraph using stackcollapse.py script.

Flame graph example, generated from `echo_server` sample:

![graph](https://github.com/zephyrproject-rtos/zephyr/assets/168683522/03a10ea9-497a-4c4e-a7e1-a70422f96317)

Kconfig optioins:
- `CONFIG_PERF` enables perf subsystem in compilation.
- `CONFIG_PERF_BUFFER_SIZE` specifies size of buffer, where samples are stored.

Kconfig requirements:
- `CONFIG_THREAD_STACK_INFO=y` provides `start` and `size` thread stack values.
- `CONFIG_SMP=n` SMP support is not implemented yet.
- `CONFIG_FRAME_POINTER=y` provide frame pointer.
- `CONFIG_SHELL=y` needs because subsystem provides shell commands.
- `CONFIG_RISCV=y` || `CONFIG_X86=y` - subsystem implemented only for x86 and riscv yet.

cfriedt commented on doc/services/debugging/perf.rst:
```diff
@@ -0,0 +1,91 @@
+
+   .. code-block:: bash
+
+      ./FlameGraph/flamegraph.pl perf_buf.foolded > graph.svg
```
```suggestion
      ./FlameGraph/flamegraph.pl perf_buf.folded > graph.svg
```

KushnerovMikhail replied:
Thanks for the comment. Corrected it.

cfriedt commented on doc/services/debugging/perf.rst:
```diff
@@ -0,0 +1,91 @@
+
+   .. code-block:: bash
+
+      python zephyr/scripts/perf/stackcollapse.py perf_buf <build_dir>/zephyr/zephyr.elf > perf_buf.foolded
```
```suggestion
      python zephyr/scripts/perf/stackcollapse.py perf_buf <build_dir>/zephyr/zephyr.elf > perf_buf.folded
```

cfriedt reviewed (dismissed):
✨.Beautiful ✨.

nashif reviewed (changes requested):
Please split the commit into multiple commits, i.e. 1 commit per arch, docs, adding the subsystem code and finally the sample.
Also, perf itself does not qualify as standalone subsystem. I suggest putting this under subsys/profiling/

Commit:
profiling: Add perf tool

Add profiling subsystem.

Add perf util based on periodic stack unwinding. Perf from Linux
was taken as a reference.

The operation of module is based on frame pointer usage and saving
registers during interruption handling.
The unwinding function stay in timer as expiry functioin so is called
during interruption handling. Thus the function have access to saved
registers (program counter and frame pointer in particular) of the current
thread and use it to unwind the thread stack.
Timer starting and results printing function are made as shell commands
for conveniency.

Originally-by: Yonatan Goldschmidt <yon.goldschmidt@gmail.com>
Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

Commit:
arch: riscv: Implement perf stack thrace func

Implement stack trace function for riscv arch, that get required
thread register values and unwind stack with it.

Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

Commit:
arch: x86: intel64: make perf stack thrace func

Implement stack trace function for x86_64 arch, that get required
thread register values and unwind stack with it.

Originally-by: Yonatan Goldschmidt <yon.goldschmidt@gmail.com>
Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

Commit:
scripts: profiling: Add stackcollapse script

Samples, that were obtained by profiling perf tool, can be be translated
into flamegraph using stackcollapse.py script.

Originally-by: Yonatan Goldschmidt <yon.goldschmidt@gmail.com>
Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

Commit:
doc: profiling: Add doc for perf tool

Add doc for profiling perf tool

Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

Commit:
samples: profiling: Add sample for perf tool

The operation of perf tool can be checked using this sample.

Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

ycsin commented on subsys/profiling/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# cOPYRIGHT (C) 2023 kns gROUP llc (yadro)
```
This and other similar ones

```suggestion
# Copyright (c) 2023 KNS Group LLC (YADRO)
```

KushnerovMikhail replied:
I don't know how I let this happen. Thanks for the comment. Corrected it.

ycsin commented on subsys/profiling/Kconfig:
```diff
@@ -0,0 +1,14 @@
+menuconfig PROFILING
+	bool "Profiling tools"
+	help
+		Enable profiling tools, such as perf
```
This and other Kconfig files, fix indent
```suggestion
	  Enable profiling tools, such as perf
```

KushnerovMikhail replied:
Thanks for the comment. Corrected it.

ycsin commented on subsys/profiling/perf/Kconfig:
```diff
@@ -0,0 +1,25 @@
+
+config PROFILING_PERF
+	bool "Perf support"
+	default n
```
n is already default
```suggestion
```

KushnerovMikhail replied:
Corrected it.

ycsin commented on subsys/profiling/perf/perf.c:
```diff
@@ -0,0 +1,98 @@
+		perf_data.idx += trace_length;
+	} else {
+		--perf_data.idx;
+		printf("perf buf override!\n");
```
should print to the shell, have a look at [k_timer_user_data_set](https://docs.zephyrproject.org/apidoc/latest/group__timer__apis.html#gadba1884961e790dd9c5d567de91cc7e2)

KushnerovMikhail replied:
I had to create an additional handler thread because the shell_print does not work in interrupts.
I hope I understood correctly how you wanted me to use `k_timer_user_data_set`.

ycsin replied:
Yeah I didn't know that the timer handler is called directly from an interrupt context, I managed to use a workqueue instead of a dedicated thread in https://github.com/zephyrproject-rtos/zephyr/commit/501c9093245fbc70df69ab3a07faa0055455e2dc (there are some other changes, as I was testing #73587)

ycsin commented on arch/riscv/core/perf.c:
```diff
@@ -0,0 +1,86 @@
+}
+
+/*
+ * This function use frame poiters to unwind stack and get trace of return addresses.
```
```suggestion
 * This function use frame pointers to unwind stack and get trace of return addresses.
```

KushnerovMikhail replied:
Corrected it.

ycsin commented on arch/riscv/core/perf.c:
```diff
@@ -0,0 +1,86 @@
+ * Return addresses are translated in corresponding function's names using .elf file.
+ * So we get function call trace
+ */
+size_t arch_perf_current_stack_trace(uintptr_t *buf, size_t size)
```
Any chance that this and https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/riscv/core/stacktrace.c can share some of the unwinding logics?

ycsin replied:
See #73587

ycsin reviewed (commented):
ARM64 supports stack unwinding as well. The docs can use some review from @kartben

KushnerovMikhail commented on subsys/profiling/perf/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	depends on THREAD_STACK_INFO
+	depends on !SMP
+	depends on !OMIT_FRAME_POINTER
+	depends on OVERRIDE_FRAME_POINTER_DEFAULT
```
Сhanged code to use a `CONFIG_FRAME_POINTER`.

KushnerovMikhail commented on subsys/profiling/perf/perf.c:
```diff
@@ -0,0 +1,98 @@
+	}
+}
+
+struct shell;
```
Really not necessary. Thanks for the comment.

KushnerovMikhail commented on subsys/profiling/perf/perf.c:
```diff
@@ -0,0 +1,98 @@
+
+	k_timer_start(&perf_data.timer, K_NO_WAIT, K_NSEC(period));
+
+	printf("Enabled perf\n");
```
Corrected it

KushnerovMikhail commented on subsys/profiling/perf/perf.c:
```diff
@@ -0,0 +1,98 @@
+
+int cmd_perf_print(const struct shell *sh, size_t argc, char **argv)
+{
+	printf("Perf buf length %zu\n", perf_data.idx);
```
Corrected it

KushnerovMikhail commented on subsys/profiling/perf/perf.c:
```diff
@@ -0,0 +1,98 @@
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(m_sub_perf,
+	SHELL_CMD_ARG(record, NULL, "Start recording for X ms on Y Hz", cmd_perf_record, 3, 0),
```
Your help string is really better then mine. Thanks for help

nashif commented on arch/riscv/core/perf.c:
```diff
@@ -0,0 +1,86 @@
+	 * The following lines lines do the reverse things to get ra, ip anf fp
+	 * from thread stack
+	 */
+	const z_arch_esf_t * const esf =
```
this was recently deprecated, can you please update? @ycsin fyi

ycsin replied:
> this was recently deprecated, can you please update? @ycsin fyi

I think it makes sense for this subsys to use a generic arch level API provided in #73587

KushnerovMikhail replied:
Fixed deprecated code

ycsin replied:
I think you are right, similar handling exists in Linux's implementation

Commit:
arch: x86: ia32: Implement perf stack thrace func

Implement stack trace function for x86_32 arch, that get required
thread register values and unwind stack with it.

Signed-off-by: Mikhail Kushnerov <m.kushnerov@yadro.com>

KushnerovMikhail commented:
Added implementation for x86_32. For it I had to replace saving %edi with %ebp in interruption context saving code. Seems like it is not ruin anything.

Also I changed a little riscv stack unwinding code to avoid bug, when interruption occurs in function prologue or epilogue. The bug was that in epilogue/prologue fp is equal to pointer of previous function frame, and it looks like second from top function is missed.

![graph](https://github.com/zephyrproject-rtos/zephyr/assets/168683522/a2bd0423-f07b-4d53-bbfd-c4aef74eaf28)
\\/ 
![graph](https://github.com/zephyrproject-rtos/zephyr/assets/168683522/07101b3e-c52c-40dd-a460-071d1226f367)
You can see that on the first graph `sys_clock_cycle_get_32` is called not only from `z_impl_k_busy_wait`, that is wrong.

Rewrite subsystem with workqueue thread instead of events

ycsin commented:
> I am thinking about using API provided in #73587, but there are two issues, I have with it:
> 
> * Riscv implementation do not check case when interruption occurred in function where $ra is not saved in stack frame (it can be if function does not call any other function). Seems like it leads to undefined behaviour (in this case `frame->ra=$fp`, `frame->fp=<undefined>`).
> * In riscv context saving code is same for exceptions and interruptions so during interruption registers are saved according to `struct arch_esf`. But for x86 there is separation between context saving during interruption and exception, and registers saved during interruption does not matches with `struct arch_esf`. And perf unwind stack during interruption.

I think from the project perspective it makes sense for applications/kernel to use the same stack-unwinding function so that there's less code duplication and easier to maintain in the long run, which is what the `arch_stack_walk()` in #73587 is trying to achieve.

I can probably fix the issue that you mentioned in RISC-V. However, I do not know how stack unwinding works in x86/arm64, but I can massage their existing implementations to use the `arch_stack_walk()`, hopefully that makes it a little easier for you to update this PR to use `arch_stack_walk()`, does this work for you?

ycsin reviewed (commented):
Should add a new entry to https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml

ycsin commented on arch/x86/core/ia32/intstub.S:
```diff
@@ -112,11 +112,12 @@ SECTION_FUNC(PINNED_TEXT, _interrupt_enter)
 	 */
 
-	/* Push EDI as we will use it for scratch space.
+	/* Push EBP as we will use it for scratch space.
```
The changes to this file could probably be in another commit / PR to document this change better

teburd commented:
Now we're cooking

nashif commented:
looks good, will take another look so we can get it ready when main branch opens
